### ISSUE TITLE
[GitHubActions or travis ci] Add test that validates oem as datapackage compatible

### ISSUE THREAD

**Issue by jh-RLI** [OpenEnergyPlatform/oemetadata#49]
I think it would be good if a[ datapackage validation](https://github.com/frictionlessdata/frictionless-py/blob/master/tests/validate/test_schema.py) is added as CI rule to the repository. The OpenEnergyMetadata repo is practically the datapackage.json and therefore should be always compatible to datapackage. This [guide ](https://framework.frictionlessdata.io/docs/guides/quick-start/)explains how to get started with frictionless datapackage.

Also add datapackage dependency to requirements.

**Comment by jh-RLI** [comment]
FYI @henhuy

**Comment by henhuy** [comment]
I'm actually implementing this in modex repo as possbility to upload datapackages to the OEP.
Still in progress. Will share link once it's working...

**Comment by jh-RLI** [comment]
the validation is already successfully used when uploading with the oedatamodel_api. We could adopt the functionality to validate the oemeta data automatically in the first step. (By extending the CI pipeline travis/gthub actions).

From the findings with the oedatamodel_api, the schema must first be adapted so that all fields of the oemetadata strign can be successfully validated. Currently only the fields of the resources are tested. This also means that the oemetadata string is not conform with the frictionless datapackage format respectively with the required datapackage.json file.

**Comment by jh-RLI** [comment]
Can be done using omi version 1 (is about to be released today)

**Comment by Cpprentice** [comment]
I tried to understand what this issue is about and how I could contribute. However, it is a little unclear to me what the goal is.

It seems it is possible to validate if a metadata JSON object is a valid datapackage using https://datapackage-validator.datist.io/. Though I don't know if that is available as a code solution somewhere - I think I remember to have seen json-schemas for the datapackage spec that could be used for that.

Frictionless also provides data validation to compare the resource schema specification of a dataset with the actual data in a csv file. This seems to be a basic feature and has an own section in the documentation https://framework.frictionlessdata.io/docs/guides/validating-data.html

There is also the concept of implementing checks and checklists which can then be used to validate a data package: https://github.com/frictionlessdata/frictionless-py/blob/399efe2d139944c985984193834e26104e765270/frictionless/package/package.py#L473

Since this was intended to be done with a CI step, I was wondering whether it is about the checking of the current release of OEMetadata actually is conformant to the datapackage spec. I guess with the above tools this would only work by comparing the example.json against the schema validation of frictionless data package?

@jh-RLI Can you shed some light on what of the above is intended with this issue?

**Comment by jh-RLI** [comment]
Yea you guessed it correctly. We discussed two things:

- Always validate oemetadata spec as frictionless datapackage (I think i and also @Ludee already worked on this)
- Using OMI as wrapper for some frictionless functionality to offer basic operations. We already have the data validation and other features integrated in omi.

The CI integration in this repo can use a example.json and frictionless-py to validate. Then we can use some csv data to integrate a data validation test for the metadata  etc. OMI should provide the functionality and already implements tests.

**Comment by Cpprentice** [comment]
I would derive the following tasks from your answer:

- Add a new test case to https://github.com/OpenEnergyPlatform/oemetadata/blob/develop/test/oemetadata/latest/test_example.py that does also validate against the datapackage spec.
- Implement this check as a new validation feature in OMI (Either by using the frictionless.validate function or the https://datapackage.org/profiles/2.0/datapackage.json schema file)
- Create a test_data.csv that is conformant to the current schema in the example.json and also validate that
- Add OMI as a test dependency to OEMetadata

Optionally we could also add invalid data - but I think that should be tested in OMI. The main purpose here is to check if any changes to the standard have an effect on the existing data validation and the schema is a valid datapackage according to the current spec.

If that is what you imagine with this I can start coding and add pull requests.

**Comment by jh-RLI** [comment]
Yes reads correctly. But first we should check again what we already have becuase some tasks have been worked on without ref. to this issue.

I think most of the required functionality is already available in OMI (see the [validation module](https://github.com/OpenEnergyPlatform/omi/blob/dev/src/omi/validation.py) and the [tests](https://github.com/OpenEnergyPlatform/omi/tree/dev/tests).

And we already have this test to check if metadata is a frictionless Package:
https://github.com/OpenEnergyPlatform/oemetadata/blob/01cd679c5e0b89315e7d2dc8fad93d5863d922db/test/oemetadata/v2/v21/test_example.py#L57


As OMI already has tests similar to what you suggested/we want, we could use them as inspiration. Having a "invalid data" test additionally is a good idea.

**Comment by Cpprentice** [comment]
I started updating the OEMetadata tests to get everything. However, it seems to me that most tests can never fail, because they just print an error message but the raised exception is caught.

Is that intentional or can I rework the tests in such a way that will fail the test if e.g. the example can not be validated by the schema?

**Comment by jh-RLI** [comment]
I See what you mean. I Did’t pay good attention when I reviewed the new Tests:)

It would be good if they fail and also show sone readable error text as well.

**Comment by Cpprentice** [comment]
Okay understood - I will update the tests for 2.1 accordingly.
We discussed before that the validation should be implemented in OMI so we could use it for the tests of OEMetadata as well. However, that creates a cyclic dependency between OMI and OEMetadata (and we would need to update both for each OEMetadata version).

My suggestion would be to move all the metadata validation code to OEMetadata directly and to import that in OMI.
OMI would then additionally contain all the stuff about data validation and validation against the OEP directly.

In OEMetadata this would mean we have a either a new file `oemetadata/validation.py` or integrate the functions directly into `oemetadata/__init__.py`


Is that a feasible strategy? And do you prefer a `validation.py` over `__init__.py`?

**Comment by jh-RLI** [comment]
In this case, I would incorporate the validation function directly into the test module as a helper. I would like to avoid providing validation functionality in oemetadata and omi – users might later wish to use this instead of the omi functionality. By integrating this functionality directly into the test, we do need to ensure that it remains synchronised with what omi does, but at least it is self-contained within the test module. 

Otherwise, I would suggest removing the validation from ‘omi’ and moving it to ‘oemetadata’, but this is not in line with the current separation of concerns.
oemetadata: specification only
omi: tooling

**Comment by Cpprentice** [comment]
Okay I can also include the helpers directly in the respective tests.

But I would include a helpers.py file directly in the test folder to keep the tests DRY.
Putting the helpers in the test folder does not include them in the dist package so no "tooling" in the package that gets shipped.

Question is though how we ensure the testing functionality stays in sync with OMI.
I guess I would split the validation in OMI as mentioned before in a metadata only validation part and the stuff also checking data or comparing with the OEP.

We could then do an occasional manual copy from the metadata validation file from omi to the oemetadata test folder.
Not sure if we should even implement this as some kind of CI step though.

**Comment by jh-RLI** [comment]
Yes seems good!

In my word:
I would have to look again - Tests are scoped to versions arent they? So we would have to add the validation for each version again. So we will have to touch the test and can then make sure the validation is in sync - for older version of the validation availabe at that time should still be okay. Does this makes sence? :)

Agree, it is okay to just include the schema against example validation. Other tests are only relevant to the tooling.

**Comment by Cpprentice** [comment]
Regarding scoping of the tests to a version:
Yes as far as I can tell each version has a new set of test files.
So I guess we can put the helpers in each version folder and just copy in the latest file from omi whenever we create a new release of oemetadata.
So even though that is not DRY, we do not need to worry about maintaining them as those would only ever get edited if we need to patch a bug.

Additionaly, OMI contains some base functionality to get a certain schema or example based on a version string.
Considering the mentioned separation of concern between omi and oemetadata:

I would argue that getting a specific oemetadata version from a string, which of course is a functionality that goes beyond the pure specification, still makes sense to be included in oemetadata directly.
OMI currently declares the `MetadataSpecification` class as a quick helper to directly access schema, template and example (and in the future example with modules).
I think we should move that including all the needed helper functions to `oemetadata.__init__.py`.

What do you think?

**Comment by Cpprentice** [comment]
I realized that OMI also performs license validation against a set of allowed license names.
That's included in its own python module - which makes sense to separate imho.
Though we either copy that over as well or we say we don't test the license for the example and template.
It is already implemented with a flag to disable that check.

**Comment by jh-RLI** [comment]
> So even though that is not DRY, we do not need to worry about maintaining them as those would only ever get edited if we need to patch a bug.

That’s what I thought too. Then I looked deeper into test design and it seems more important not to tie the tests too closely to the code modules. Given that we also have this ‘workflow’ of one test suite per version, I think this is the better solution, even if it doesn’t quite adhere to the DRY principle.


> I would argue that getting a specific oemetadata version from a string, which of course is a functionality that goes beyond the pure specification, still makes sense to be included in oemetadata directly.
OMI currently declares the MetadataSpecification class as a quick helper to directly access schema, template and example (and in the future example with modules).
I think we should move that including all the needed helper functions to oemetadata.__init__.py.
What do you think?

At present, this change would affect several implementations in various tools and on the OEP. Whilst a redesign would be feasible, I ampacked with other tasks so i cant help here - and i think this would be required here. I would also like to give this change a little more thought. In principle, however, I agree that this specific functionality can be scoped to the specification itself.

> I realized that OMI also performs license validation against a set of allowed license names. That's included in its own python module - which makes sense to separate imho. Though we either copy that over as well or we say we don't test the license for the example and template. It is already implemented with a flag to disable that check.

I think the license check must not existis in oemetadata - Yes i added the flag because of this :D
